src/main/scala/org/broadinstitute/dsde/firecloud/service/UserService.scala

```diff
@@ -110,4 +123,86 @@ class UserService(rawlsDAO: RawlsDAO, thurloeDAO: ThurloeDAO, userToken: UserInf
       case Success(_) => getTerraPreference(userToken)
     }
   }
-}
+
+  private def getWord(iL: Long, wordList: List[String]): String = {
+    val modIndex: Int = (math.abs(iL) % (wordList.length - 1)).toInt
+    wordList(modIndex)
+  }
+
+  private def writeAnonymousGroup(userToken: UserInfo, anonymousGroupName: String): Future[PerRequestMessage] = {
+    val kvpsToUpdate = Map(
+      UserService.AnonymousGroupKey -> anonymousGroupName
+    )
+
+    logger.info(s"${userToken.userEmail} (${userToken.id}) setting anonymousGroup to $anonymousGroupName")
+
+    thurloeDAO.saveKeyValues(userToken, kvpsToUpdate) flatMap {
+      case Failure(exception) => Future(RequestCompleteWithErrorReport(StatusCodes.InternalServerError,
+        "could not save Anonymous Group", exception))
+      case Success(_) => {
+        val futureAllKeys: Future[ProfileWrapper] = getAllKeysFromThurloe(userToken)
+        futureAllKeys map { keys => RequestComplete(keys)}
+      }
+    }
+  }
+
+  private def getAnonymousGroup: String = {
+    // randomly generate the anonymousGroupName, which follows format: terra-user-adjective-noun-endOfUUID
+    val anonymousGroupUUID: UUID = UUID.randomUUID()
+    val anonymousGroupName: String = (FireCloudConfig.FireCloud.supportPrefix
+      + getWord(anonymousGroupUUID.getMostSignificantBits(), UserService.randomAdjectiveList) + "-"
+      + getWord(anonymousGroupUUID.getLeastSignificantBits(), UserService.randomNounList) + "-"
+      + anonymousGroupUUID.toString().split("-")(4))
+    anonymousGroupName
+  }
+
+  private def getAllKeysFromThurloe(userToken: UserInfo): Future[ProfileWrapper] = {
+    thurloeDAO.getAllKVPs(userToken.id, userToken) map { // .getAllKVPs returns Option[ProfileWrapper]
+      case None => ProfileWrapper(userToken.id, List())
+      case Some(wrapper) => {
+        wrapper
+      }
+    }
+  }
+
+  private def callCreateGoogleGroup(keys: ProfileWrapper, anonymousGroupName: String): Future[PerRequestMessage] = {
+    // define userEmail to add to google Group - check first for contactEmail, otherwise use user's login email
+    val userEmail = getProfileValue(keys, UserService.ContactEmailKey) match {
+      case None | Some("") => userToken.userEmail // if the contactEmail key doesn't exist OR (|) if there's no contactEmail set in the profile
+      case Some(contactEmail) => contactEmail // if there is a non-empty value set for contactEmail, we assume contactEmail is a valid email
+    }
+
+    googleServicesDAO.createGoogleGroup(anonymousGroupName, userEmail) match { // this will return "" if group creation is not successful
```

Comment:
Generic feedback, I think some of the inline could be grouped into method documentation instead, but that's just personal preference. The fact that it's there is great!